fix: using tombstones to account for rapid deletion

shawkins opened the pull request:
closes: #2314

If during a creation the events for both the add and the deletion occur before the resource is being added to the temporary cache, the temporary cache moves backwards in a way that may not resolve itself.

The approach here is to use tombstones to track deletions. The benefit of this approach is that it keeps the logic encapsulated to the TemporaryResourceCache. The drawback is that we're only guessing about how long to retain the tombstones.

The other approach would look more similar to event recording, which is notify the cache that a create is happening prior to when it starts so that deletions / tombstones can gathered for that specific resource.

csviri commented on operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java:
```diff
@@ -60,7 +63,23 @@ protected boolean removeEldestEntry(java.util.Map.Entry<String, Boolean> eldest)
+  synchronized void onEvent(T resource, boolean unknownState) {
+    if (!tombstones.isEmpty()) {
+      var iter = tombstones.entrySet().iterator();
+      var entry = iter.next();
```
Not sure if understand this part, it always removes one random event from the tombstone if expired? Didn't you mean to iterate over all events?

shawkins replied:
No, this is to amortorize the expiration logic - you could of course decide on a greater number than 1.

csviri replied:
Maybe desribe in there as a comment too, that is intentional, since I will at least naivly stop on this every time I open it :)

metacosm replied:
I would like to see this as a comment, indeed, as well.

csviri commented on operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java:
```diff
@@ -38,8 +38,11 @@ public class TemporaryResourceCache<T extends HasMetadata> {
 
   private static final Logger log = LoggerFactory.getLogger(TemporaryResourceCache.class);
   private static final int MAX_RESOURCE_VERSIONS = 256;
+  private static final int TOMBSTONE_RETENTION_TIME = 120000;
```
Wondering about this value, if high enough.

shawkins replied:
It is certianly possible there could be an extreme scenario where the client is configured for a bunch of exponential retries - I'm not sure there is a hard cap on the amount of time in that case. We can keep things for longer, or you can go back the other direction of telling the cache prior to the create that you are intending to create the given resource.

csviri commented on operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java:
```diff
@@ -84,14 +103,25 @@ public synchronized void putResource(T newResource, String previousResourceVersi
+            newResource.getMetadata().getUid(), resourceId);
+        return;
+      }
+      moveAhead = true;
```
THis seems to be correct, but would be good to document why is this here, isn't is covered already by the condition on line 120-121 ? or just an optimization?

shawkins replied:
I'll add a comment, it's just breaking apart the checks that were being performed.

csviri replied:
thank you!

metacosm reviewed (changes requested):
LGTM apart from missing comment explaining the amortization of cache emptying logic

shawkins commented:
@metacosm @metacosm updated the code with further comments and consolidated the secondary caching concerns to a single class. From what I can tell it will take around 170 MB to track up to 1000000 deletions. Also updated the base retention time to 10 minutes, rather than 2. Based upon seeing what the retained heap size of this could be, I also updated the knownResourceVersions cache to be able to store more entries.
